<html>
<head>
  <title>异常总结</title>
  <basefont face="微软雅黑" size="2" />
  <meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
  <meta name="exporter-version" content="YXBJ Windows/604043 (zh-CN, DDL); Windows/10.0.0 (Win64); EDAMVersion=V2;"/>
  <style>
    body, td {
      font-family: 微软雅黑;
      font-size: 10pt;
    }
  </style>
</head>
<body>
<a name="558"/>
<h1>异常总结</h1>

<div>
<span><div><div style="margin: 10px auto; padding: 0px; list-style-type: none; list-style-image: none; text-indent: 0px; letter-spacing: normal; orphans: 2; text-align: start; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: #ffffff;"><div style="font-size: 14px;"><span style="font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial; font-family: gotham, helvetica, arial, sans-serif; font-size: 14px;">异常可以在某些东西出错时自动引发。为了引发异常有两种方式：</span></div></div><div style="background-color: #f5f5f5; border: 1px solid #cccccc; padding: 5px; overflow: auto; margin: 5px 0px; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; font-size: 12px !important;"><div style="margin: 0px; white-space: pre-wrap; word-wrap: break-word; padding: 0px; list-style-type: none; list-style-image: none; font-size: 12px !important;"><div><span style="font-size: 12px; color: #800080; font-family: 'Courier New'; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">1</span><span style="font-size: 12px; color: #000000; font-family: 'Courier New'; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">.使用一个类，应该是Excepion的子类（使用类时，程序会自动创建类的一个实例）</span></div><div><span style="font-size: 12px; color: #800080; font-family: 'Courier New'; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">2</span><span style="color: #000000; font-family: 'Courier New'; font-size: 12px; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">.实例参数调用raise语句。</span></div></div></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; font-weight: bold;">处理异常</span>：try语句：</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">1.try-except：</span></div><ul><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">可以有多个except语句分别处理相应的异常；</span></div></li><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">也可以将若干异常放在元组中统一处理；</span></div></li><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">捕获所有异常可以使用Exception或使用excpet裸语句</span></div></li><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">except Exception as e:</span></div></li></ul><ul><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">e为Exception的实例</span></div></li><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">如果要捕获所有异常，可以直接捕获Exception</span></div></li></ul><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">2.try-except-finally:可以有不止一个except子句；不管能否捕获异常都会执行finally子句</span></div><div><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px;">3.try-finally：只允许检测异常并做一些必要的清理工作，没有任何异常处理设施；当finally中的</span><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; font-style: italic;">所有代码都执行完之后会继续向上一层引发</span><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px;">异常。无论try块是否发生异常都可以利用try/finally中的finally执行清理工作</span></div><div><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px;">4.try-except-else-finally:except和else都是可选的，except有一个或多个；如果不执行except(try没有发生异常)，就会执行else,</span><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; text-decoration: underline;">有了else块可以尽量缩减try块内的代码量</span><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px;">，try中应该只抛出预期异常的语句，使其更加易读</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">一个最终注意点：如果finally中的代码引发另一个异常或由于return，break,continue而终止，原来的异常将丢失而无法重新引发</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">5.以下代码：</span></div><div style="box-sizing: border-box; padding: 8px; font-family: Monaco, Menlo, Consolas, 'Courier New', monospace; font-size: 12px; color: #333333; border-radius: 4px; background-color: #fbfaf8; border: 1px solid rgba(0, 0, 0, 0.15); -en-codeblock: true;"><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;">def func():</span></div><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;"><span>    try:</span><br clear="none"/></span></div><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;"><span><span>    </span><span>    return 'from try'</span><br clear="none"/></span></span></div><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;"><span><span>    finally:</span><br clear="none"/></span></span></div><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;"><span><span>    </span><span>    return 'from finally'</span><br clear="none"/></span></span></div><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;"><span>&gt;&gt;&gt;func()</span></span></div><div><span style="font-size: 9pt; color: #333333; font-family: Monaco;"><span>'from finally'</span></span></div></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span>    说明：</span><br clear="none"/></span></div><ul><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span>当在try...finally语句中try执行return, break, continue后，finally子句依然会执行</span></span></div></li><li><div>函数的返回值由最后的return语句决定，由于finally子句一定会执行，所以finally子句中的return将始终是最后执行的语句</div></li></ul><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; font-weight: bold;">触发异常</span>：</span></div><ul><li><div><span style="font-size: 17px;"><span style="font-family: 'Noto Serif', serif; font-size: 17px; color: #333333;">raise Exception(</span><span style="font-family: 'Noto Serif', serif; font-size: 17px; color: #ff0000;">*args</span><span style="font-family: 'Noto Serif', serif; font-size: 17px; color: #333333;">)</span></span></div></li></ul><ul><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif; line-height: 26.7143px;">基类的参数类型是*args，所有的参数都放在一个元组中,可以添加任何参数</span></div></li></ul><ul><li><div><span style="font-family: georgia, palatina, serif;">断言：assert expression</span></div></li></ul><ul><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">如果表达式为真不做任何操作，为False会触发异常：AssertionError</span></div></li><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">等价于：if __debug__:\n if not expression:raise AssertionError</span></div></li><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">貌似不能带参数</span></div></li></ul><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; font-weight: bold;">捕获异常</span>：</span></div><ul><li><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">捕获异常后抛出另外的异常：raise from：先抛出之前的异常，再抛出此次异常：</span></div><div>try:</div><div>     int('N/A')</div><div>except ValueError as e:</div><div>     raise RuntimeError('A parsing error occurred') from e</div></li><li><div><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px;">重新抛出被捕获的异常：在except语句块中使用单独的raise，很常见的用法是在捕获所有异常的处理器中重新抛出——</span><span style="color: #ff0000; font-family: 'Noto Serif', serif; font-size: 17px;">为什么要这样做</span><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px;">？</span></div><div>try:</div><div>     int('N/A')</div><div>except ValueError:</div><div>     print(&quot;Didn't work&quot;)</div><div>     raise</div></li></ul><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; font-weight: bold;">自定义异常</span>：Python3 要求自定义的异常必须继承 Exception 类。内建的所有异常也都是继承自这个类。因此，我们只需要捕获这个类的实例，就可以捕获所有的异常。</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">    定义新的异常类重写了__init__方法，确保总是用所有传递过来的参数调用Exception.__init__</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;"><span style="color: #333333; font-family: 'Noto Serif', serif; font-size: 17px; font-weight: bold;">内建异常类</span>：</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">BaseException</span></div><div><span style="font-size: 17px; color: #333333; font-family: 'Noto Serif', serif;">列表解析：</span></div><div><span style="font-size: 14px; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: #ffffff; float: none; color: #000000; font-family: Arial; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">在一个序列的值上应用一个任意表达式，将其结果收集到一个新的列表中并返回。它的基本形式是一个方括号里面包含一个for语句对一个iterable对象迭代</span></div><div><span style="font-size: 14px; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: #ffffff; float: none;"><br clear="none"/></span></div><div><span style="font-size: 14px; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: #ffffff; float: none; color: #000000; font-family: Arial; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">NotImplementedError: 继承自RunTimeError，用来定义父类的的抽象函数需要子类继承，可抛出该异常</span></div><div><span style="font-size: 14px; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: #ffffff; float: none;"><br clear="none"/></span></div><div><span style="font-size: 14px; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: #ffffff; float: none; color: #000000; font-family: georgia, palatina, serif; font-style: normal; font-variant-caps: normal; font-variant-ligatures: normal; font-weight: normal; text-decoration-color: initial; text-decoration-style: initial;">除了assert语句之外，python还提供内置的只读变量 __debug__， 除非解释器运行在最优模式中(通过使用-O选项指定)，否则它的值为True。在运行额外的错误检查过程时，程序可能在需要时检查这个变量。__debug__变量的底层实现在解释器中经过了优化，因为实际上没有包含if语句本身的额外控制流逻辑。如果python运行在普通模式中，if __debug__语句下的语句就会内联到没有if语句本身的程序中。在最优模式中，if __debug__ 语句以及所有相关语句都将从程序中完全删除。</span></div></div></span>
</div></body></html> 